### ISSUE TITLE
Portcontrol over SNMPv3

### ISSUE THREAD

**Issue by jonnycastaway** [netdisco/netdisco#616]
Hi,
we use snmpv3 only in our network and Netdisco is getting all necessary data from the switches with snmpv3. But now we want to set adminstatus of ports up and down with netdisco. But the gets queued all the time and the switchport doesn't changed state. I set read and write to true in deployment.yaml but nothing. When i'm doing this with SnmpB (an snmp test tool) all is working well.
How Netdisco try to switch ports with portcontrol? Is ist possible to doing that with snmpv3 and Netdisco?

**Comment by inphobia** [comment]
interesting question.

first off: not all snmp::info classes support setting values, which netdisco & snmp::info versions are you using and what is the device you're trying to access?

to see why netdisco is having issues can you try the following command? keep in mind to change the device ip and port name:
`~/bin/netdisco-do portcontrol -DI -d 192.0.2.1 -p FastEthernet0/1 -e down`

**Comment by philfin42** [comment]
I'm having this issue as well operating on Cisco 9400. All other read functions work. Other platforms using the same snmp v3 rw creds work as well. 

App::Netdisco | 2.42.10
SNMP::Info | 3.68

[9146] 2019-07-22 13:34:16  info App::Netdisco version 2.042010 loaded.
[9146] 2019-07-22 13:34:17  info portcontrol: [hidden ip] started at Mon Jul 22 09:34:17 2019
[9146] 2019-07-22 13:34:17 debug portcontrol: running with timeout 600s
[9146] 2019-07-22 13:34:17 debug => running workers for phase: check
[9146] 2019-07-22 13:34:17 debug -> run worker check/_base_/0
[9146] 2019-07-22 13:34:17 debug PortControl is able to run
[9146] 2019-07-22 13:34:17 debug => running workers for phase: main
[9146] 2019-07-22 13:34:17 debug -> run worker main/_base_/100
[9146] 2019-07-22 13:34:17 debug snmp writer cache warm: [hidden ip]
[9146] 2019-07-22 13:34:17 debug [hidden] try_connect with ver: 3, class: SNMP::Info::Layer3::Cisco, comm: <hidden>
SNMP::Info::_global load_location : SNMPv2-MIB::sysLocation.0 : .1.3.6.1.2.1.1.6.0
SNMP::Info::_set set_location.0 (.1.3.6.1.2.1.1.6.0) = 
SNMP::Info::_set noAccess at /home/netdisco/perl5/lib/perl5/App/Netdisco/Transport/SNMP.pm line 266.
[9146] 2019-07-22 13:34:17 debug [hidden ip] try_connect with ver: 3, class: SNMP::Info, comm: <hidden>
SNMP::Info::_global load_location : SNMPv2-MIB::sysLocation.0 : .1.3.6.1.2.1.1.6.0
SNMP::Info::_set set_location.0 (.1.3.6.1.2.1.1.6.0) = 
SNMP::Info::_set noAccess at /home/netdisco/perl5/lib/perl5/App/Netdisco/Transport/SNMP.pm line 266.
[9146] 2019-07-22 13:34:17 debug failed to connect to (hidden ip) to update up_admin
[9146] 2019-07-22 13:34:17  info portcontrol: finished at Mon Jul 22 09:34:17 2019
[9146] 2019-07-22 13:34:17  info portcontrol: status defer: failed to connect to (hidden ip) to update up_admin

**Comment by inphobia** [comment]
strange stuff, any vrf or access groups you're using? acls perhaps, remotely authenticated users?

has this ever worked in the past btw, we did make a few changes so make rfc1213 load in a different way, which is in that code path.

**Comment by StraussJ11** [comment]
I had the same issue. i discovered the network using SNMPv1 community strings (only read), and after that, i switched to SNMPv3 read/write. After the change, i was not immediately able to make changes to ports. A "discover all" job resolved the issue for me. The regular auto-discover did not help, i really had to click the button (Admin->Discover all)

**Comment by inphobia** [comment]
also please test:
https://github.com/netdisco/netdisco/wiki/Troubleshooting#change-the-snmp-community-string-for-a-device

**Comment by inphobia** [comment]
so `failed to connect to (IP ADDRESS) to update up_admin` means that the writer_for() call failed somehow.

according to the info in your previous message it might be that the _try_write() function somehow is failing.

from what i can make of the code it tries to read the device location & then write it back to see if it can snmp write. so, 2 questions:

1. do you have privileges to write to the location oid of the device?
1. does the device currently have an snmp location or is it blank? can you try all these in this order?
   * ~/bin/netdisco-do show -d IP -DI -e load_location
   * ~/bin/netdisco-do show -d IP -DI -e location
   * ~/bin/netdisco-do show -d IP -DI -e sysLocation

**Comment by philfin42** [comment]
The location was blank. I ran a debug on the switch at the same time I was trying to bring the port down from netdisco. 
SrParseV3SnmpMessage: No matching Engine ID.
SrParseV3SnmpMessage: Failed.
SrDoSnmp: authentication failure, Unknown Engine ID

Which lead me to Google, and then lead me to having to add:
snmp-server view groupname iso included

Once I added that to the switch, netdisco port control is working! Thanks for your help and assistance.
